### PR TITLE
Added dynamic bloom filter, added murmur3 hash

### PR DESCRIPTION
**Pull request by igoreus**

### CODE REVIEW

File: src/BloomFilter/BloomFilterAbstract.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+     *
+     * @return int
+     */
+    protected function hash($value, $index)
+    {
+        return $this->hash->hash($value . $index) % $this->bitSize;
+    }
+
+    /**
+     * m = ceil((n * log(p)) / log(1.0 / (pow(2.0, log(2.0)))));
```
Comment (nicolopignarocket) [inline src/BloomFilter/BloomFilterAbstract.php]: comments not needed?

File: src/BloomFilter/BloomFilterAbstract.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+     * @param float $falsePositiveProbability
+     * @return int
+     */
+    private function getOptimalBitSize($setSize, $falsePositiveProbability = 0.001)
+    {
+        return (int) round((($setSize * log($falsePositiveProbability)) / pow(log(2), 2)) * -1);
+    }
+
+    /**
+     * k = round(log(2.0) * m / n);
```
Comment (nicolopignarocket) [inline src/BloomFilter/BloomFilterAbstract.php:96]: comments not needed?

File: example/example.php [hunk 3]
```diff
@@ -28,19 +29,22 @@ function microtime_float()
...
 ];
 
 echo 'Messages: ' . count($messages) . "\n";
 
+/**
+ * @var  string $storage
+ * @var \RocketLabs\BloomFilter\BloomFilterInterface $filter
+ */
 foreach ($filters as $storage => $filter) {
+    $redis->del('bloom_filter');
```
Comment (fstaoe) [inline example/example.php:45]: Why should the redis key be deleted in every iteration?

File: example/example.php [hunk 4]
```diff
@@ -3,6 +3,8 @@
 
 use RocketLabs\BloomFilter\Persist\Redis;
 use RocketLabs\BloomFilter\BloomFilter;
+use \RocketLabs\BloomFilter\DynamicBloomFilter;
```
Comment (nicolopignarocket) [inline example/example.php]: remove \

**Comment by nicolopignarocket** [comment]
general reminder that libraries should never throw native PHP exceptions, but always custom defined ones

File: src/BloomFilter/BloomFilterAbstract.php [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    /**
+     * @param string $value
+     * @param int $index
+     *
+     * @return int
+     */
+    protected function hash($value, $index)
+    {
+        return $this->hash->hash($value . $index) % $this->bitSize;
```
Comment (nicolopignarocket) [inline src/BloomFilter/BloomFilterAbstract.php]: hash->hash, eye to eye https://www.youtube.com/watch?v=nozewQJ4hyw

File: src/BloomFilter/BloomFilterAbstract.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        $bloom->init($setSize, $falsePositiveProbability);
+
+        return $bloom;
+    }
+
+    /**
+     * @param PersisterInterface $persister
+     * @param HashInterface $hash
+     */
+    public function __construct(PersisterInterface $persister, HashInterface $hash)
```
Comment (fstaoe) [inline src/BloomFilter/BloomFilterAbstract.php]: Would it not be better to have this method and the init with private visibility?

File: src/BloomFilter/BloomFilterAbstract.php [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+    /**
+     * @param PersisterInterface $persister
+     * @param HashInterface $hash
+     * @param int $setSize
+     * @param float $falsePositiveProbability
+     *
+     * @return BloomFilterInterface
+     */
+    public static function createFromSetSize(
```
Comment (fstaoe) [inline src/BloomFilter/BloomFilterAbstract.php]: In my opinion, would be more clear to have the factory method as abstract and each child class has to implement it... having to call the abstract class constructor and execute the code which is on the current init method.

**Pull request merged**